chrono-tz/src/timezone_impl.rs

```diff
@@ -403,3 +404,69 @@ impl TimeZone for Tz {
         TzOffset::new(*self, timespans.get(index))
     }
 }
+
+/// Represents the information of a gap.
+pub struct GapInfo {
+    /// When available it contains information about the beginning of the gap.
+    ///
+    /// The time represents the first instant in which the gap starts.
+    /// This means that it is the first instant that when used with [`TimeZone::from_local_datetime`]
+    /// it will return [`LocalResult::None`].
+    ///
+    /// The offset represents the offset of the first instant before the gap.
+    pub begin: Option<(NaiveDateTime, TzOffset)>,
+    /// When available it contains the first instant after the gap.
+    pub end: Option<DateTime<Tz>>,
+}
+
+impl Tz {
+    /// Returns information about a gap.
+    ///
+    /// It returns `None` if `local` is not in a gap for the current timezone.
+    ///
+    /// If `local` is at the limits of the known timestamps the fields `begin` or `end` in
+    /// [`GapInfo`] will be `None`.
+    pub fn gap_info_from_local_datetime(&self, local: &NaiveDateTime) -> Option<GapInfo> {
+        let timestamp = local.and_utc().timestamp();
+        let timespans = self.timespans();
+        let index = binary_search(0, timespans.len(), |i| {
+            timespans.local_span(i).cmp(timestamp)
+        });
+
+        match index {
```

Comment:
Please use an early return for the top-level `None` case, maybe with let-else:

```rust
let Err(end_idx) = index else {
    return None;
};
```

For `begin` and `end`, suggest using `match`:

```rust
let begin = match end_idx {
    0 => None,
    _ => {
        ..
    }
};
```

Reply:
> Please use an early return for the top-level `None` case, maybe with let-else:
> 
> ```rust
> let Err(end_idx) = index else {
>     return None;
> };
> ```

Thanks this make the code much better!

> 
> For `begin` and `end`, suggest using `match`:
> 
> ```rust
> let begin = match end_idx {
>     0 => None,
>     _ => {
>         ..
>     }
> };
> ```

I kept it with `if/else` because for the `end` case a match is not possible because we are matching against a variable (unless you want to use a guard) and I didn't like having `match` for `begin` and `if/else` for `end`. Again happy to change.

Reply:
Instead of `match` or `if/else` what do you think about
```rust
let begin = (end_idx > 0).then_some(end_idx - 1).map(|start_idx| {
    timespans
        .local_span(start_idx)
        .end
       // ...
});

let end = (end_idx < timestamps.len()).then_some(end_idx).map(|end_idx| {
    timespans
        .local_span(end_idx)
        .begin
       // ...
});
```

Reply:
I think I'd prefer using match for both, with a guard.

Reply:
Done